sugondat-shim/src/cli.rs

```diff
@@ -153,64 +148,104 @@ pub mod query {
+                BlockRef::Hash(h) => write!(f, "0x{}", hex::encode(&h[..])),
+            }
+        }
+    }
+
+    impl std::str::FromStr for BlockRef {
+        type Err = String;
+
+        fn from_str(input: &str) -> Result<Self, Self::Err> {
+            if input == "best" {
+                return Ok(BlockRef::Best);
+            }
+
+            if let Some(hash) = decode_hash(input)? {
+                return Ok(BlockRef::Hash(hash));
+            }
 
-        use super::SugondatRpcParams;
-
-        /// A reference to a block to query.
-        #[derive(Debug, Clone)]
-        pub enum BlockRef {
-            /// The current best finalized block known by the node.
-            Best,
-            /// The number of the block to query.
-            Number(u64),
-            /// The hex-encoded hash of the block to query, prefixed with "0x".
-            Hash([u8; 32]),
+            if let Ok(n) = input.parse::<u64>() {
+                Ok(BlockRef::Number(n))
+            } else {
+                Err(format!("parse error. see `--help`"))
+            }
         }
+    }
 
-        impl std::fmt::Display for BlockRef {
-            fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
-                match *self {
-                    BlockRef::Best => write!(f, "best"),
-                    BlockRef::Number(n) => write!(f, "{}", n),
-                    BlockRef::Hash(h) => write!(f, "0x{}", hex::encode(&h[..])),
-                }
+    fn decode_hash(input: &str) -> Result<Option<[u8; 32]>, String> {
+        if let Some(s) = input.strip_prefix("0x") {
+            let bytes =
+                hex::decode(s).map_err(|_| "Invalid parameter: not hex encoded".to_owned())?;
+
+            let mut hash = [0u8; 32];
+            if bytes.len() != 32 {
+                return Err("Invalid parameter: hash not 32 bytes".to_owned());
             }
+
+            hash.copy_from_slice(&bytes[..]);
+            Ok(Some(hash))
+        } else {
+            Ok(None)
         }
+    }
 
-        impl std::str::FromStr for BlockRef {
-            type Err = String;
+    pub mod blob {
+        use clap::Args;
 
-            fn from_str(input: &str) -> Result<Self, Self::Err> {
-                if input == "best" {
-                    return Ok(BlockRef::Best);
-                }
+        use super::{BlockRef, SugondatRpcParams};
 
-                if let Some(s) = input.strip_prefix("0x") {
-                    let bytes = hex::decode(s)
-                        .map_err(|_| "Invalid parameter: not hex encoded".to_owned())?;
+        #[derive(Debug, Args)]
+        pub struct Params {
+            #[clap(flatten)]
+            pub rpc: SugondatRpcParams,
 
-                    let mut hash = [0u8; 32];
-                    if bytes.len() != 32 {
-                        return Err("Invalid parameter: hash not 32 bytes".to_owned());
-                    }
+            /// The block to query information about.
```

Comment:
I think this needs a better comment: it just feels like this functionality is about querying a block whereas we query a blob at a specific block. Here block is just only part of identifier/specifier of a location, and not the thing we are querying.